test_mavros/include/tests/offboard_control.h

```diff
@@ -0,0 +1,470 @@
+#include <sitl_test/test_type.h>
+
+#include <geometry_msgs/Point.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <geometry_msgs/TwistStamped.h>
+
+namespace testtype {
+/**
+ * @brief Offboard controller tester
+ *
+ * Tests offboard position, velocity and acceleration control
+ *
+ */
+class OffboardControl {
+public:
+	OffboardControl() :
+		nh_sp("~"),
+		mode("position"),
+		shape("square")
+	{ };
+
+	/* -*- main routine -*- */
+
+	void spin(int argc, char *argv[]) {
+		/**
+		 * @brief Setpoint control mode selector
+		 *
+		 * Available modes:
+		 * - position
+		 * - velocity
+		 * - acceleration
+		 */
+		nh_sp.param<std::string>("mode", mode, "position");
+
+		/**
+		 * @brief Setpoint path shape selector
+		 *
+		 * Available shapes:
+		 * - square
+		 * - circle
+		 * - eight
+		 * - ellipse (in 3D space)
+		 */
+		nh_sp.param<std::string>("shape", shape, "square");
+
+		ros::Rate loop_rate(10);
+
+		ROS_INFO("SITL Test: Offboard control test running!");
+
+		if (mode.compare("position") == 0) {
+			ROS_INFO("Position control mode selected.");
+			if (shape.compare("square") == 0) {
+				ROS_INFO("Test option: square-shaped path...");
+				square_path_motion(loop_rate);
+			}
+			else if (shape.compare("circle") == 0) {
+				ROS_INFO("Test option: circle-shaped path...");
+				circle_path_motion(loop_rate);
+			}
+			else if (shape.compare("eight") == 0) {
+				ROS_INFO("Test option: eight-shaped path...");
+				eight_path_motion(loop_rate);
+			}
+			else if (shape.compare("ellipse") == 0) {
+				ROS_INFO("Test option: ellipse-shaped path...");
+				ellipse_path_motion(loop_rate);
+			}
+		}
+		else if (mode.compare("velocity") == 0) {
+			ROS_INFO("Velocity control mode selected.");
+			if (shape.compare("square") == 0) {
+				ROS_INFO("Test option: square-shaped path...");
+				square_path_motion(loop_rate);
+			}
+			else if (shape.compare("circle") == 0) {
+				ROS_INFO("Test option: circle-shaped path...");
+				circle_path_motion(loop_rate);
+			}
+			else if (shape.compare("eight") == 0) {
+				ROS_INFO("Test option: eight-shaped path...");
+				eight_path_motion(loop_rate);
+			}
+			else if (shape.compare("ellipse") == 0) {
+				ROS_INFO("Test option: ellipse-shaped path...");
+				ellipse_path_motion(loop_rate);
+			}
+		}
+		else if (mode.compare("acceleration") == 0) {
+			ROS_INFO("Velocity control mode selected.");
+			//TODO
+			return;
+		}
+	}
+
+private:
+	ros::NodeHandle nh_sp;
+	ros::Publisher local_pos_sp_pub;
+	ros::Publisher vel_sp_pub;
+	ros::Subscriber local_pos_sub;
+
+	std::string mode;
+	std::string shape;
+
+	geometry_msgs::PoseStamped localpos;
+	geometry_msgs::PoseStamped ps;
+
+	geometry_msgs::TwistStamped vs;
+
+	/* -*- alias -*- */
+    #define X  ps.pose.position.x
+    #define Y  ps.pose.position.y
+    #define Z  ps.pose.position.z
+    #define VX  vs.twist.linear.x
+    #define VY  vs.twist.linear.y
+    #define VZ  vs.twist.linear.z
+    #define sp_X  pos_setpoint().x
```

Comment:
Or you can use Eigen for math, and then convert it to msg (`eigen_conversions.h`).